packages/dev/serializers/src/USDZ/USDZSerializer.ts

```diff
@@ -0,0 +1,732 @@
+        const array = [];
+        for (const uuid in materials) {
+            const material = materials[uuid];
+
+            array.push(await USDZExport._BuildMaterial(material, textures, quickLookCompatible));
+        }
+
+        return `def "Materials"
+    {
+    ${array.join("")}
+    }
+    `;
+    }
+
+    private static _ChannelIndexMap: { [key: string]: number } = {
+        r: 0,
+        g: 1,
+        b: 2,
+        a: 3,
+    };
+
+    private static async _CreateDynamicTextureFromChannel(texture: Texture, channel: string, scene: Scene): Promise<DynamicTexture> {
+        const dt = new DynamicTexture("dt", { width: texture.getSize().width, height: texture.getSize().height }, scene, false);
+        const ctx = dt.getContext();
+        const img = (await texture.readPixels()) as any;
+        const data = (ctx as any).createImageData(texture.getSize().width, texture.getSize().height);
+        const count = texture.getSize().width * texture.getSize().height * 4;
+
+        for (let i = 0; i < count; i += 4) {
+            const value = channel === "r" ? 255 - img[i + USDZExport._ChannelIndexMap[channel]] : img[i + USDZExport._ChannelIndexMap[channel]];
+            data.data[i] = value;
+            data.data[i + 1] = value;
+            data.data[i + 2] = value;
+            data.data[i + 3] = 255;
+        }
+
+        ctx.putImageData(data, 0, 0);
+        dt.update(texture.invertY);
+        return dt;
+    }
+
+    private static async _BreakApartMetallicRoughnessAo(texture: Texture, scene: Scene): Promise<{ ao: DynamicTexture; roughness: DynamicTexture; metallic: DynamicTexture }> {
+        const ao = await USDZExport._CreateDynamicTextureFromChannel(texture, "r", scene);
+        ao.name = "ao";
+        const roughness = await USDZExport._CreateDynamicTextureFromChannel(texture, "g", scene);
+        roughness.name = "roughness";
+        const metallic = await USDZExport._CreateDynamicTextureFromChannel(texture, "b", scene);
+        metallic.name = "metallic";
+        return { ao, roughness, metallic };
+    }
+
+    private static async _BuildMaterial(_material: Material, textures: { [id: string]: Texture }, quickLookCompatible = false) {
+        // https://graphics.pixar.com/usd/docs/UsdPreviewSurface-Proposal.html
+        const pad = "			";
+        const inputs = [];
+        const samplers = [];
+        const material = _material as PBRMaterial;
+        const scene = material.getScene();
+
+        //TODO: rescope this whole function.
+        function buildTexture(texture: Texture, mapType: string, color: Color3 | undefined = undefined) {
+            const id = texture.uniqueId + "_" + texture.invertY;
+            textures[id] = texture;
+
+            const uv = texture.coordinatesIndex > 0 ? `st${texture.coordinatesIndex}` : `st`;
+
+            /* TODO CHECK THIS */

```

Comment:
What needs to be checked? As I can see it, you translated the code from Threejs, so that should be good?

Reply:
should be, but I'm not sure how to test it though.

Reply:
Ohh that todo was for the wrappings to be checked.